language/evm/move-to-yul/tests/ControlStructures.exp.apply-cfg-to-scf

```diff
@@ -0,0 +1,309 @@
+                            // label L3
+                            // $t7 := 1
+                            $t7 := 1
+                            // $t0 := +($t0, $t7)
+                            x := $AddU64(x, $t7)
+                            // goto L6
+                            $block4 := 7
+                        }
+                        case 6 {
+                            // label L5
+                            // $t8 := 2
+                            $t8 := 2
+                            // $t0 := -($t0, $t8)
+                            x := $Sub(x, $t8)
+                            // goto L6
+                            $block4 := 7
+                        }
+                        case 7 {
+                            // label L6
+                            // goto L7
+                            $block4 := 4
+                        }
+                    }
+                }
+                case 3 {
+                    // label L2
+                    // return ()
+                    leave
+                }
+            }
+
+            function A2_M_g(x) {
+                let $t1, $t2, $t3, $t4, $t5
+                let $block := 4
+                switch $block
+                case 4 {
+                    let $block4 := 4
+                    for {} true {} {
+                        switch $block4
+                        case 3 {
+                            // label L2
+                            // $t4 := 0
+                            $t4 := 0
+                            // $t5 := ==($t0, $t4)
+                            $t5 := $Eq(x, $t4)
+                            // if ($t5) goto L4 else goto L3
+                            switch $t5
+                            case 0  { $block4 := 4 }
+                            default { $block4 := 5 }
+                        }
+                        case 4 {
+                            // label L3
+                            // $t1 := 1
+                            $t1 := 1
+                            // $t2 := >=($t0, $t1)
+                            $t2 := $GtEq(x, $t1)
+                            // if ($t2) goto L0 else goto L2
+                            switch $t2
+                            case 0  { $block4 := 3 }
+                            default { $block4 := 2 }
+                        }
+                    }
+                }
+                case 5 {
+                    // label L4
+                    // return ()
+                    leave
+                }
+            }
+
+            function A2_M_h1(x) -> $result {
+                let tmp_$1, $t2, $t3, $t4, $t5
+                let $block := 4
+                switch $block
+                case 4 {
+                    // $t2 := 0
+                    $t2 := 0
+                    // $t3 := >($t0, $t2)
+                    $t3 := $Gt(x, $t2)
+                    // if ($t3) goto L0 else goto L2
+                    switch $t3
+                    case 0  { $block := 3 }
```

Comment:
I think the general pattern to reduce this to an if-then-else is as follows. Given (after topsort):

```
L1:
    B1
    if c goto L2 else L3
L2:
    B2
    goto L4
L3:
    B3        
L4:
```

... becomes:

```
B1
switch c
case 0 { B3 }
default { B2 }
```

So one would need to identify these patterns n the code.

IMO one would best do this by introducing a temporary SCG data structures which represents if-then-else and possibly also loops.

Reply:
> @wrwg Thanks for the comment! I was a bit stuck and thus checked out how Cheer implemented it, I found that it had more granular "tokens", or "bytecodes" in our repo iiuc, https://github.com/leaningtech/cheerp-compiler/blob/30a05221c34f418374dd8b8149d63f0917477d34/llvm/lib/CheerpWriter/CheerpWriter.cpp#L5009-L5028
> 
> so I am wondering do we need a separate set of bytecodes other than those in "stackless_bytecodes.rs", so that we can generate if-else and while based on the more granular ones instead?

I'm not sure what you mean with finer granularity. We have a rather regular representation of bytecode and we certainly won't change any of this. The code you are pointing to seems unrelated to the problem at hand, it is just a compiler from some IR to a text form.

> 
> > a temporary SCG data structures which represents if-then-else and possibly also loops.
> 
> can you elaborate a bit how SCG here would help detect the pattern?

I'm not talking about how to detect the pattern, but how to represent the result. However, thinking about the result may  guide you also how to do this:

```
enum SCG {
   Block{ start_offs: usize, end_offs: usize } // slice of code without any branches
   If{ cond: TempIndex, if_true: Box<SCG>, if_false: Box<SCG> }
   Loop{ body: SCG } // perhaps need a new opcode for continue and break
```


> 
> Thanks a lot!

Reply:
@wrwg thanks for the explanation!

> The code you are pointing to seems unrelated to the problem at hand, it is just a compiler from some IR to a text form.

I was thinking of generating text forms with blocks, now the text forms are generated based on the related bytecode, more specifically "branch" for control flows. Contrary to this, in the code pointer, it has a notion of "token", which has different tokens for "if", "while" etc.  With the "branch" we can directly generate the switch style of text codes while to generate if-else, I guessed we would need some similar notion of "tokens" there?

>  how to represent the result

I see, this makes sense!

Reply:
> @wrwg thanks for the explanation!
> 
> > The code you are pointing to seems unrelated to the problem at hand, it is just a compiler from some IR to a text form.
> 
> I was thinking of generating text forms with blocks, now the text forms are generated based on the related bytecode, more specifically "branch" for control flows. 

That is what I mean with this SCG data structure. You generate this one, then change the code in generator.rs to generate Yul from the SCG.

Contrary to this, in the code pointer, it has a notion of "token", which has different tokens for "if", "while" etc. With the "branch" we can directly generate the switch style of text codes while to generate if-else, I guessed we would need some similar notion of "tokens" there?
> 
> > how to represent the result
> 
> I see, this makes sense!